src/frontend/src/lib/utils/token.utils.ts

```diff
@@ -131,3 +139,104 @@ export const mapTokenUi = ({
 		$exchanges
 	})
 });
+
+/**
+ * Type guard to check if a token is of type RequiredTokenWithLinkedData.
+ * This checks whether the token has a twinTokenSymbol field and ensures that it is a string.
+ *
+ * @param token - The token object to be checked.
+ * @returns A boolean indicating whether the token is of type RequiredTokenWithLinkedData.
+ */
+export const isRequiredTokenWithLinkedData = (token: Token): token is RequiredTokenWithLinkedData =>
+	'twinTokenSymbol' in token && typeof token.twinTokenSymbol === 'string';
+
+/**
+ * Type guard to check if an object is of type TokenGroupUi.
+ *
+ * @param tokenUiOrGroupUi - The object to check.
+ * @returns A boolean indicating whether the object is a TokenGroupUi.
+ */
+export const isTokenGroupUi = (tokenUiOrGroupUi: TokenUiOrGroupUi): tokenUiOrGroupUi is TokenGroupUi =>
+	typeof tokenUiOrGroupUi === 'object' &&
+	!isNullish(tokenUiOrGroupUi) &&
+	'header' in tokenUiOrGroupUi &&
+	typeof (tokenUiOrGroupUi).header === 'object' &&
+	'tokens' in tokenUiOrGroupUi;
+
+/**
+ * Factory function to create a TokenGroupUi based on the provided tokens and network details.
+ * This function creates a group header and adds both the native token and the twin token to the group's tokens array.
+ *
+ * @param nativeToken - The native token used for the group, typically the original token or the one from the selected network.
+ * @param twinToken - The twin token to be grouped with the native token, usually representing the same asset on a different network.
+ *
+ * @returns A TokenGroupUi object that includes a header with network and symbol information and contains both the native and twin tokens.
+ */
+const createTokenGroup = ({
+	nativeToken,
+	twinToken
+}: {
+	nativeToken: TokenUi;
+	twinToken: TokenUi;
+}): TokenGroupUi => ({
+	header: {
+		name: nativeToken.network.name,
+		symbol: `${nativeToken.symbol}, ${twinToken.symbol}`,
+		decimals: nativeToken.decimals,
+		icon: nativeToken.icon
+	},
+	nativeNetwork: nativeToken.network,
+	tokens: [nativeToken, twinToken]
+});
+
+/**
+ * Function to create a list of TokenUiOrGroupUi by grouping tokens with matching twinTokenSymbol.
+ * The group is placed in the position where the first token of the group was found.
+ * Tokens with no twin remain as individual tokens in their original position.
+ *
+ * @param tokens - The list of TokenUi objects to group. Each token may or may not have a twinTokenSymbol.
+ *                 Tokens with a twinTokenSymbol are grouped together.
+ *
+ * @returns A new list where tokens with twinTokenSymbols are grouped into a TokenGroupUi,
+ *          and tokens without twins remain in their original place.
+ *          The group replaces the first token of the group in the list.
+ */
+export const groupTokensByTwin = (tokens: TokenUi[]): TokenUiOrGroupUi[] =>
```

Comment:
right now this function suppose that the ckTokens comes always after their twin counterparty in the order of `tokens` input (that is generally the case, I suppose?): if a ckToken appears first in the list, what happens?

Furthermore, it considers only groups of two by two: for example, if, for whatever reason we have ETH, ckETH and ETH-BNB, that refers all to ETH, they will not be grouped. That is ok for now since it is not needed.

However, in the interest of time and just to track it, may you please add a TODO here to mention both issues above?

Reply:
We have changed the order of the tokens countless times, so for future-proofing, it might be better to implement logic that makes no assumptions about the ordering.

Regarding the second point, we discussed this week that SepoliaETH might eventually be grouped with ETH. Therefore, I believe it would be future-proof not to assume that there are only two tokens per group.

Reply:
Given that there were already too many comments in the PR, I suggest we add a TODO and address those questions in follow-up PRs. I mean, not something we schedule, but right after we merge this one.

Reply:
Yes, that bug was introduced. I reverted it to the easy implementation: ed145f4d. I will ping Stefan to do a tech debt ticket to reimplement the function in an ideologically correct way.

Reply:
No need to add it to debt, it's a bug we want to fix before using it. In another PR but, we gonna fix that before any usage.

Reply:
https://dfinity.atlassian.net/browse/OISY-142

Reply:
If i may give a suggestion on implementation: I would avoid creating a list of mixed tokens and grouped tokens.
I think it is to return a list of just groups, and in some of those groups there is only a single token.

Then you handle differently single-element groups from multiple-elements groups, but you do it elsewhere.

Doing so will simplify the grouping function quite a bit, I think, and it would use only one `reduce` with no additional loops.